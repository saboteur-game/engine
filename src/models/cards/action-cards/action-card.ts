import Card from "../card";
import CardParameters from "../card-parameters";
import { Pojo } from "../../../utils";

class ActionCard extends Card {
  parameters: CardParameters | undefined;

  play(parameters: CardParameters): ActionCard {
    super.play(parameters);
    this.parameters = parameters;
    return this;
  }

  toJSON(): Pojo {
    return {
      ...super.toJSON(),
      parameters: this.parameters,
    };
  }
}

export default ActionCard;
