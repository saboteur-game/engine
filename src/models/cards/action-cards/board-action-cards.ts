import { ActionCard } from ".";
import CardParameters from "../card-parameters";

export class BoardActionCard extends ActionCard {
  play(parameters: CardParameters): BoardActionCard {
    return super.play(parameters);
  }
}

export class MapActionCard extends BoardActionCard {}

export class RockfallActionCard extends BoardActionCard {}
