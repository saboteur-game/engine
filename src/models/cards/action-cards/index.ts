import ActionCard from "./action-card";
import {
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
} from "./board-action-cards";
import {
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
} from "./tool-action-cards";

export {
  ActionCard,
  BoardActionCard,
  MapActionCard,
  RockfallActionCard,
  ToolActionCard,
  BreakActionCard,
  RepairActionCard,
};
