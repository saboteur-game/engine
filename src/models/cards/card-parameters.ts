import { Tools } from "../../constants";
import Position from "../position";
import Player from "../player";

export interface IBoardCardParameters {
  position: Position;
}

export interface IToolActionCardParameters {
  player: Player;
  appliedTo: Tools;
}

type CardParameters = IToolActionCardParameters | IBoardCardParameters;

export default CardParameters;
