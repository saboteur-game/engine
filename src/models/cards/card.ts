import { generateId, Pojo } from "../../utils";
import CardParameters from "./card-parameters";

export enum Status {
  unused = "unused",
  played = "played",
  discarded = "discarded",
}

class Card {
  id: string;
  status: Status;

  constructor() {
    this.id = generateId();
    this.status = Status.unused;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  play(parameters: CardParameters): Card {
    return this;
  }

  setDiscarded(): void {
    // TODO: Check if the state was unused and throw if it was anything else
    this.status = Status.discarded;
  }

  setPlayed(): void {
    // TODO: Check if the state was unused or played and throw if it was anything else
    this.status = Status.played;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      status: this.status,
    };
  }
}

export default Card;
