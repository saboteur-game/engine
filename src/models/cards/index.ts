import { shuffle } from "../../utils";
import { Tools } from "../../constants";
import Card, { Sides } from "./card";
import {
  StartPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
  PassageCard,
  DeadendCard,
  PathCard,
} from "./path-cards";
import {
  MapActionCard,
  RockfallActionCard,
  BreakActionCard,
  RepairActionCard,
  ActionCard,
} from "./action-cards";

const multiply = (length: number, getValue: () => Card) =>
  new Array(length).fill(undefined).map(() => getValue());
const randomBoolean = () => Math.random() < 0.5;

/*
 * Saboteur cards
 * - 44 path cards
 *  -- 1 starting card
 *  -- 3 finishing cards
 *    --- 1 gold [Connected, 1-2-3-4]
 *    --- 2 rock
 *       ---- 1 [Connected 1-2]
 *       ---- 1 [Connected 1-4]
 *  -- 40 tunnel cards
 *    --- 31 passage cards
 *       ---- 3 passage [Connected 2-4]
 *       ---- 4 passage [Connected 1-3]
 *       ---- 4 passage [Connected 1-4]
 *       ---- 5 passage [Connected 1-2]
 *       ---- 5 passage [Connected 1-2-4]
 *       ---- 5 passage [Connected 1-2-3]
 *       ---- 5 passage [Connected 1-2-3-4]
 *    --- 9 dead-end cards
 *       ---- 1 dead-end [Connected 1]
 *       ---- 1 dead-end [Connected 2]
 *       ---- 1 dead-end [Connected 1-2]
 *       ---- 1 dead-end [Connected 1-3]
 *       ---- 1 dead-end [Connected 1-4]
 *       ---- 1 dead-end [Connected 2-4]
 *       ---- 1 dead-end [Connected 1-2-3]
 *       ---- 1 dead-end [Connected 1-2-4]
 *       ---- 1 dead-end [Connected 1-2-3-4]
 * - 27 action cards
 *  -- 6 map cards
 *  -- 3 rock-fall cards
 *  -- 9 break tool cards
 *    --- 3 pick
 *    --- 3 lamp
 *    --- 3 wagon
 *  -- 9 repair tool cards
 *    --- 2 pick
 *    --- 2 lamp
 *    --- 2 wagon
 *    --- 1 pick and lamp
 *    --- 1 pick and wagon
 *    --- 1 lamp and wagon
 *
 * Notes: Card connections
 *
 *        (1)
 *     +-------+
 *     |       |
 * (4) |       | (2)
 *     |       |
 *     +-------+
 *        (3)
 */

export interface IPlacedCards {
  start: StartPathCard;
  gold: GoldFinishPathCard;
  rock1: RockFinishPathCard;
  rock2: RockFinishPathCard;
}

export const getPlacedCards = (): IPlacedCards => {
  const start = new StartPathCard();
  const gold = new GoldFinishPathCard();
  const rock1 = new RockFinishPathCard([Sides.top, Sides.right]);
  const rock2 = new RockFinishPathCard([Sides.top, Sides.left]);

  [start, gold, rock1, rock2].forEach((card) => card.setPlayed());

  return { start, gold, rock1, rock2 };
};

export const getShuffledDeck = (): Array<ActionCard | PathCard> => {
  const pathCards = [
    ...multiply(
      3,
      () => new PassageCard([Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      4,
      () => new PassageCard([Sides.top, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () => new PassageCard([Sides.top, Sides.right], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.left], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard([Sides.top, Sides.right, Sides.bottom], randomBoolean())
    ),
    ...multiply(
      5,
      () =>
        new PassageCard(
          [Sides.top, Sides.right, Sides.bottom, Sides.left],
          randomBoolean()
        )
    ),
    new DeadendCard([Sides.top], randomBoolean()),
    new DeadendCard([Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right], randomBoolean()),
    new DeadendCard([Sides.top, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.left], randomBoolean()),
    new DeadendCard([Sides.right, Sides.left], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.bottom], randomBoolean()),
    new DeadendCard([Sides.top, Sides.right, Sides.left], randomBoolean()),
    new DeadendCard(
      [Sides.top, Sides.right, Sides.bottom, Sides.left],
      randomBoolean()
    ),
  ] as Array<PathCard>;

  const actionCards = [
    ...multiply(6, () => new MapActionCard()),
    ...multiply(3, () => new RockfallActionCard()),
    ...multiply(3, () => new BreakActionCard(Tools.pick)),
    ...multiply(3, () => new BreakActionCard(Tools.lamp)),
    ...multiply(3, () => new BreakActionCard(Tools.wagon)),
    ...multiply(2, () => new RepairActionCard(Tools.pick)),
    ...multiply(2, () => new RepairActionCard(Tools.lamp)),
    ...multiply(2, () => new RepairActionCard(Tools.wagon)),
    new RepairActionCard([Tools.pick, Tools.lamp]),
    new RepairActionCard([Tools.pick, Tools.wagon]),
    new RepairActionCard([Tools.lamp, Tools.wagon]),
  ] as Array<ActionCard>;

  return shuffle(
    ([] as Array<ActionCard | PathCard>).concat(pathCards, actionCards)
  );
};
