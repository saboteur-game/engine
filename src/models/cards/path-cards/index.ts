import PathCard from "./path-card";
import { TunnelCard, PassageCard, DeadendCard } from "./tunnel-cards";
import {
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
} from "./boundary-cards";

export {
  PathCard,
  TunnelCard,
  PassageCard,
  DeadendCard,
  StartPathCard,
  FinishPathCard,
  GoldFinishPathCard,
  RockFinishPathCard,
};
