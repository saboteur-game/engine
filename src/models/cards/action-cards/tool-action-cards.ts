import { Tools } from "../../../constants";
import { ActionCard } from ".";
import { Pojo } from "../../../utils";
import { IToolActionCardParameters } from "../card-parameters";

export class ToolActionCard extends ActionCard {
  appliesTo: Tools[];

  constructor(appliesTo: Tools | Tools[]) {
    super();
    this.appliesTo = ([] as Tools[]).concat(appliesTo);
  }

  play(parameters: IToolActionCardParameters): ActionCard {
    if (!this.appliesTo.includes(parameters.appliedTo)) {
      throw new Error(
        `Invalid parameter provided. Cannot apply to ${parameters.appliedTo}`
      );
    }
    return super.play(parameters);
  }

  toJSON(): Pojo {
    return {
      ...super.toJSON(),
      appliesTo: this.appliesTo,
    };
  }
}

export class BreakActionCard extends ToolActionCard {}

export class RepairActionCard extends ToolActionCard {}
