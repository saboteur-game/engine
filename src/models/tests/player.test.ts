import Player from "../player";
import { MapActionCard } from "../cards/action-cards";
import { Tools } from "../../constants";
import { FinishPathCard, RockFinishPathCard } from "../cards/path-cards";
import Position from "../position";
import { Sides } from "../cards/card";

const PLAY_POSITION = new Position(0, 1);

describe("Player", () => {
  let card: MapActionCard;
  let player: Player;

  beforeEach(() => {
    card = new MapActionCard();
    player = new Player("Alice", 35);
  });

  it("can be serialized", () => {
    expect(player.toJS()).toMatchSnapshot();
  });

  it("will be provided random age if unspecified", () => {
    const { age } = new Player("Bob");
    expect(age).toBeGreaterThanOrEqual(8);
    expect(age).toBeLessThanOrEqual(90);
  });

  describe("addToHand", () => {
    it("adds the card to the players hand", () => {
      expect(player.getHandCardCount()).toBe(0);
      player.addToHand(card);
      expect(player.getHandCardCount()).toBe(1);
    });
  });

  describe("removeFromHand", () => {
    describe("when the card is in the players hand", () => {
      beforeEach(() => {
        player.addToHand(card);
      });

      it("removes the card from the players hand", () => {
        expect(player.getHandCardCount()).toBe(1);
        player.removeFromHand(card.id);
        expect(player.getHandCardCount()).toBe(0);
      });
    });

    describe("when the card is not in the players hand", () => {
      it("throws exception", () => {
        expect(() => player.removeFromHand(card.id)).toThrow(
          "Cannot remove card which isn't in players hand"
        );
      });
    });
  });

  describe("addToHand", () => {
    describe("when there are no cards in the players hand", () => {
      it("returns empty array", () => {
        expect(player.getHand()).toEqual([]);
      });
    });

    describe("when there are cards in the players hand", () => {
      beforeEach(() => {
        player.addToHand(card);
      });

      it("returns populated array", () => {
        expect(player.getHand()).toMatchSnapshot();
      });
    });
  });

  describe("playCard", () => {
    beforeEach(() => {
      player.addToHand(card);
    });

    describe("when card is not in the players hand", () => {
      it("throws exception", () => {
        expect(() =>
          player.playCard("test-card-id-123", { position: PLAY_POSITION })
        ).toThrow("Cannot play card which isn't in players hand");
      });
    });

    describe("when card is in the players hand", () => {
      it("returns played card with parameters", () => {
        expect(card.parameters).toBe(undefined);
        expect(player.playCard(card.id, { position: PLAY_POSITION })).toBe(
          card
        );
        expect(card.parameters).toEqual({ position: PLAY_POSITION });
      });
    });
  });

  describe("discardCard", () => {
    beforeEach(() => {
      player.addToHand(card);
    });

    describe("when card is not in the players hand", () => {
      it("throws exception", () => {
        expect(() => player.discardCard("test-card-id-123")).toThrow(
          "Cannot discard card which isn't in players hand"
        );
      });
    });

    describe("when card is in the players hand", () => {
      it("returns discarded card and removes it from the players hand", () => {
        expect(player.getHandCardCount()).toBe(1);
        expect(player.discardCard(card.id)).toBe(card);
        expect(player.getHandCardCount()).toBe(0);
      });
    });
  });

  describe("breakTool", () => {
    describe("when tool is not already broken", () => {
      it("sets the corresponding tool to off", () => {
        expect(player.getTools()).toMatchInlineSnapshot(`
          Object {
            "lamp": true,
            "pick": true,
            "wagon": true,
          }
        `);
        player.breakTool(Tools.lamp);
        expect(player.getTools()).toMatchInlineSnapshot(`
          Object {
            "lamp": false,
            "pick": true,
            "wagon": true,
          }
        `);
      });
    });

    describe("when tool is already broken", () => {
      beforeEach(() => {
        player.breakTool(Tools.lamp);
      });

      it("throw exception", () => {
        expect(() => player.breakTool(Tools.lamp)).toThrow(
          "Tool lamp is already broken"
        );
      });
    });
  });

  describe("repairTool", () => {
    describe("when tool is not already repaired", () => {
      beforeEach(() => {
        player.breakTool(Tools.lamp);
      });

      it("sets the corresponding tool to off", () => {
        expect(player.getTools()).toMatchInlineSnapshot(`
          Object {
            "lamp": false,
            "pick": true,
            "wagon": true,
          }
        `);
        player.repairTool(Tools.lamp);
        expect(player.getTools()).toMatchInlineSnapshot(`
          Object {
            "lamp": true,
            "pick": true,
            "wagon": true,
          }
        `);
      });
    });

    describe("when tool is already repaired", () => {
      it("throw exception", () => {
        expect(() => player.repairTool(Tools.lamp)).toThrow(
          "Tool lamp does not need repaired"
        );
      });
    });
  });

  describe("setup", () => {
    let finishCard: RockFinishPathCard;

    beforeEach(() => {
      finishCard = new RockFinishPathCard([Sides.top]);
      player.viewFinishCard(finishCard);
      player.addToHand(card);
      player.breakTool(Tools.pick);
      player.breakTool(Tools.lamp);
      player.breakTool(Tools.wagon);
    });

    it("resets the players hand", () => {
      expect(player.getHand()).toEqual([card]);
      player.setup(false);
      expect(player.getHand()).toEqual([]);
    });

    it("resets broken tools", () => {
      expect(player.getTools()).toMatchInlineSnapshot(`
        Object {
          "lamp": false,
          "pick": false,
          "wagon": false,
        }
      `);
      player.setup(false);
      expect(player.getTools()).toMatchInlineSnapshot(`
        Object {
          "lamp": true,
          "pick": true,
          "wagon": true,
        }
      `);
    });

    it("resets any viewed finish cards", () => {
      expect(player.getViewedFinishCards()).toEqual([finishCard]);
      player.setup(false);
      expect(player.getViewedFinishCards()).toEqual([]);
    });

    it("sets player to saboteur when provided true", () => {
      player.setup(true);
      expect(player.isSaboteur).toBe(true);
    });

    it("sets player to gold digger when provided false", () => {
      player.setup(false);
      expect(player.isSaboteur).toBe(false);
    });
  });

  describe("viewFinishCard", () => {
    describe("when viewing general card", () => {
      it("throws exception", () => {
        expect(() => player.viewFinishCard(card as FinishPathCard)).toThrow(
          "Invalid finish card provided"
        );
        expect(player.getViewedFinishCards().length).toBe(0);
      });
    });

    describe("when viewing finish card", () => {
      let finishCard: RockFinishPathCard;

      beforeEach(() => {
        finishCard = new RockFinishPathCard([
          Sides.top,
          Sides.right,
          Sides.bottom,
          Sides.left,
        ]);
      });

      it("adds to the list of viewed cards", () => {
        expect(player.getViewedFinishCards().length).toBe(0);
        player.viewFinishCard(finishCard);
        expect(player.getViewedFinishCards().length).toBe(1);
        expect(player.getViewedFinishCards()[0]).toBe(finishCard);
      });
    });
  });
});
