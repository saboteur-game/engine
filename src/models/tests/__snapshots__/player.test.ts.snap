// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player addToHand when there are cards in the players hand returns populated array 1`] = `
Array [
  MapActionCard {
    "id": "test-id-0",
    "status": "unused",
  },
]
`;

exports[`Player can be serialized 1`] = `
Object {
  "age": 35,
  "hand": Object {},
  "id": "test-id-1",
  "isSaboteur": false,
  "name": "Alice",
  "viewedFinishCards": Object {},
}
`;
